Return account's asset balances

jsidorenko opened the pull request:
This method solves the problem of fetching the account's asset balances, now instead of sending 100+ RPC requests it would be possible to send just one.

Cumulus Companion: https://github.com/paritytech/cumulus/pull/2180

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2371944

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2371948

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2371949

gilescope commented on frame/assets/src/functions.rs:
```diff
@@ -924,4 +924,11 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 	}
+
+	/// Used by the RPC service to provide account balances.
+	pub fn account_balances(_instance: u8, account: T::AccountId) -> Vec<(T::AssetId, T::Balance)> {
```
_instance unused seems a bit odd?

jsidorenko replied:
```suggestion
	pub fn account_balances(_: u8, account: T::AccountId) -> Vec<(T::AssetId, T::Balance)> {
```

gilescope commented on frame/assets/runtime-api/src/lib.rs:
```diff
@@ -0,0 +1,35 @@
+		AssetBalance: Codec,
+		AssetId: Codec
+	{
+		/// Returns the list of asset id and non-zero balance the AccountId has.
```
```suggestion
		/// Returns an unsorted list of asset id and non-zero balances the AccountId has.
```

bkchr replied:
Stuff is always unsorted, if not specified otherwise.

bkchr commented:
Do we really need this here in Substrate? I mean the runtime api could also just be declared for Statemint/Statemine. It could also there directly have concrete namings like `asset_x_account_balances`, `asset_y_account_balances`.

gilescope commented:
Cross-chain consistency of getting asset balances is better than bespoke naming. I'd suggest that being able to get the balances of an account is something that should come as standard.

bkchr commented:
> Cross-chain consistency of getting asset balances is better than bespoke naming. I'd suggest that being able to get the balances of an account is something that should come as standard.

While I agree with this. The current implementation here with a `instance: u8` is also not really some standard. What is the instance? The index of the pallet in the runtime? Or just the the order of the asset pallets on how they appear in the in construct runtime? This isn't really defined and also not a "standard".

In the future when we have https://forum.polkadot.network/t/wasm-view-functions/1045 we can have a standard.

I will approve the pr for now, but I request some better docs on how "instance" is defined.

bkchr commented on frame/assets/src/functions.rs:
```diff
@@ -924,4 +924,11 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 		})
 	}
+
+	/// Used by the RPC service to provide account balances.
```
```suggestion
	/// Returns all the non-zero balances for all assets of the given `account`.
```

xlc reviewed (changes requested):
The fact this assumes a chain can have multiple kinds of asset id is just super confusing and not making sense.

Yes a chain can have multiple pallet-assets instance with different AssetId type for each instance. However the chain SHOULD still have a single unified asset id. It could be 
```
enum AssetId {
  LocalAssets(u32),
  ForeignAssets(MultiLocation),
}
```
and having one pallet-assets handle local assets and another one handle foreign assets. But the API / UI / UX should never require user to specify an instance ID AND an asset id.

jsidorenko commented:
> The fact this assumes a chain can have multiple kinds of asset id is just super confusing and not making sense.

Where do you see that assumption?
`fn account_balances(instance: u8, account: AccountId) -> Vec<(u32, Balance)> {` states the AssetId will be u32 and the Balance will be of the same type regardless of the actual assets instance

jsidorenko commented:
I removed the instance param as it causes too much confusion

jsidorenko commented:
> The fact this assumes a chain can have multiple kinds of asset id is just super confusing and not making sense.
> 
> Yes a chain can have multiple pallet-assets instance with different AssetId type for each instance. However the chain SHOULD still have a single unified asset id. It could be
> 
> ```
> enum AssetId {
>   LocalAssets(u32),
>   ForeignAssets(MultiLocation),
> }
> ```
> 
> and having one pallet-assets handle local assets and another one handle foreign assets. But the API / UI / UX should never require user to specify an instance ID AND an asset id.

What would be the proper solution then?

xlc commented:
Acala for example, we have pallet-balances, orml-tokens, and ERC20 from EVM and we have a single CurrencyId for all those.

https://github.com/AcalaNetwork/Acala/blob/01f19ae1bfb0afff62a0b938c8e2d02cf5d831c5/primitives/src/currency.rs#L466

xlc commented:
But this is kind of a bad idea anyway. It could be very very expensive to iterate all the assets. This could be a DoS vector to the node. Take a look at Ethereum, how many ERC20 are there and how are you going to iterate every ERC20 on Ethereum to return a balance list?

jsidorenko commented:
@xlc In Ethereum, there are balance contracts those accept an array of contracts to check and return you up to 1k balances at once

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2382435

jsidorenko commented:
@xlc I removed the traits definition from the pallet and moved it into the node/runtime/lib file. So that, for each pallet's instance it would be possible to create a new trait/implementation/endpoint

xlc commented:
> @xlc In Ethereum, there are balance contracts those accept an array of contracts to check and return you up to 1k balances at once

It takes array of contracts, which limits the max lookups and caller needs to know the contracts upfront.
It returns up to 1k balances, which limits the max response size.

xlc commented:
I am ok with this as it is. But just saying it doesn't scale and we will need alternative solution in future.

ruseinov reviewed (approved):
looks good to me, but could potentially be expensive

kianenigma reviewed (changes requested):
I think this flies in the face of using XCM as a query language, as described [here](https://forum.polkadot.network/t/xcm-as-a-standard-for-reading-and-interacting-with-parachains/266). IMO the return type should be an XCM a collection of `MultiAsset`, not just something related to assets pallet. 

This could go in, but it is a short-sighted approach. 

The real solution is to use MultiAsset, and have a registry somewhere that maps MultiAssets to token names etc, which is also being worked on [here](https://forum.polkadot.network/t/xcm-global-asset-registry-gar-call-for-contributors/1470). 

Then, the workflow is that any chain would implement a single API, whereby a wallet sends a request, provides an account id, and you get all the `MultiAssets` associated with it. Then the wallet can continue and convert MultiAssets to something human readable. 

By extension, you can start using the same for transfers as well, as highlighted in the original post. 

All of this is much more holistic and long-lasting, than what this PR is trying to achieve.

bkontur commented:
> I think this flies in the face of using XCM as a query language, as described [here](https://forum.polkadot.network/t/xcm-as-a-standard-for-reading-and-interacting-with-parachains/266). IMO the return type should be an XCM a collection of `MultiAsset`, not just something related to assets pallet.

@kianenigma 
but MultiAsset is part of xcm, which is in polkadot repo, right?

kianenigma commented:
> > I think this flies in the face of using XCM as a query language, as described [here](https://forum.polkadot.network/t/xcm-as-a-standard-for-reading-and-interacting-with-parachains/266). IMO the return type should be an XCM a collection of `MultiAsset`, not just something related to assets pallet.
> 
> @kianenigma but MultiAsset is part of xcm, which is in polkadot repo, right?

Yes, but that's an irrelevant technical matter. You can just declare the API in the Polkadot repo.

kianenigma commented:
I am happy to remove my review, but I'd like the PR authors to also study the links that I have provided and express if they agree that the correct solution is to not only solve this at the level of assets pallet, but at the XCM layer. 

Currently, an account can have "_value bearing assets_" in balances/assets pallet, in a DEX pallet, in some liquid staking pool pallet, in a lending platform. A wallet in our ecosystem has no meaningful way to scrape all of this information as it stands now, unless if they go on and study each parachain/platform one by one. 

The XCM-centric approach would solve all of this in one go, and moves the responsibility of implementing this API to each individual chain, which is where it should logically reside.

Polkadot-Forum commented:
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/xcm-as-a-standard-for-reading-and-interacting-with-parachains/266/5

kianenigma commented on frame/assets/src/functions.rs:
```diff
@@ -924,4 +924,11 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 	}
+
+	/// Returns all the non-zero balances for all assets of the given `account`.
+	pub fn account_balances(account: T::AccountId) -> Vec<(T::AssetId, T::Balance)> {
```
```suggestion
	pub fn account_balances(account: T::AccountId, max_iters: Option<u32>) -> Vec<(T::AssetId, T::Balance)> {
```

xlc replied:
I don’t think this approach will work in general. This will need to have proper paging support to be exhaustive and this kind of API is simply impossible for ERC20 tokens implementation.

kianenigma commented on frame/assets/src/functions.rs:
```diff
@@ -924,4 +924,11 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
+
+	/// Returns all the non-zero balances for all assets of the given `account`.
+	pub fn account_balances(account: T::AccountId) -> Vec<(T::AssetId, T::Balance)> {
+		Asset::<T, I>::iter_keys()
```
```suggestion
		Asset::<T, I>::iter_keys().take(max_iters.unwrap_or(u32::MAX))
```

kianenigma commented on bin/node/runtime/src/assets_api.rs:
```diff
@@ -0,0 +1,34 @@
+use codec::Codec;
+use sp_std::vec::Vec;
+
+sp_api::decl_runtime_apis! {
```
Why is this re-defined in every single runtime in Polkadot?

bkontur replied:
there is just one common definition for all statemine/t runtimes (statemine/statemint/westmint),
but lets discuss it here: https://github.com/paritytech/cumulus/pull/2180#discussion_r1113053040

shawntabrizi commented:
I am not against this PR getting in if it solves some specific problem we need tomorrow, but yeah, the XCM route seems like the right approach. We should instead have a new XCM Query RPC, and one of the XCM queries should be exactly like this, and implementable by runtime developers to work against any pallets.

mustermeiszer commented:
> I am not against this PR getting in if it solves some specific problem we need tomorrow, but yeah, the XCM route seems like the right approach. We should instead have a new XCM Query RPC, and one of the XCM queries should be exactly like this, and implementable by runtime developers to work against any pallets.

* What is the advantage of having XCM-query-rpc over normal query-rpc?

bkontur commented:
> I am not against this PR getting in if it solves some specific problem we need tomorrow, but yeah, the XCM route seems like the right approach. We should instead have a new XCM Query RPC, and one of the XCM queries should be exactly like this, and implementable by runtime developers to work against any pallets.

@shawntabrizi 
yes, please, I also would like to know more about "XCM Query RPC", what exactly does that mean?
Does it mean something like "expose XCM-dedicated runtime api", e.g.: https://github.com/paritytech/cumulus/pull/2180/files#diff-4bcbbb9c6ce00315b062b0f075d0fb9f452e3091420b5d2234b092385c09b716R56-R64 ?

bkontur commented:
bot rebase

bkontur commented:
bot rebase

jsidorenko commented:
bot rebase

jsidorenko commented:
bot merge

Polkadot-Forum commented:
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-release-analysis-v0-9-40/2468/1
